packages/chatbot-server-mongodb-public/src/processors/formatUserMessageForGeneration.ts

```diff
@@ -0,0 +1,48 @@
+/**
+  Adds front matter to the user message text.
+ */
+import { ConversationCustomData, logger } from "mongodb-rag-core";
+
+export function formatUserMessageForGeneration(
+  userMessageText: string,
+  customData?: ConversationCustomData
+) {
+  const frontMatters: Record<string, string> = {};
+  const { origin, originCode } = customData ?? {};
+  if (typeof origin !== "string" || !origin) {
+    return userMessageText;
+  }
+  try {
+    const url = new URL(origin);
+    if (
+      url.hostname === "mongodb.com" ||
+      url.hostname.endsWith(".mongodb.com")
+    ) {
+      frontMatters.pageUrl = origin;
+    }
+  } catch (e) {
+    logger.warn(`Origin ${origin} malformed. Not using in front matter.`);
+  }
+
+  if (originCode === "VSCODE") {
+    frontMatters.client = "MongoDB VS Code plugin";
+  } else if (originCode === "GEMINI_CODE_ASSIST") {
+    frontMatters.client = "Gemini Code Assist";
+  }
+  return addFrontMatter(frontMatters, userMessageText);
+}
+
+function addFrontMatter(
```

Comment:
there's a utility func in core called updateFrontmatter() which you can use here instead of this